src/main/java/co/cask/gcp/datastore/sink/DatastoreSink.java

```diff
@@ -0,0 +1,343 @@
+import co.cask.cdap.api.data.schema.Schema.Field;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.batch.BatchRuntimeContext;
+import co.cask.cdap.etl.api.batch.BatchSink;
+import co.cask.cdap.etl.api.batch.BatchSinkContext;
+import co.cask.gcp.datastore.sink.util.SinkKeyType;
+import co.cask.gcp.datastore.util.DatastorePropertyUtil;
+import co.cask.hydrator.common.LineageRecorder;
+import com.google.cloud.Timestamp;
+import com.google.cloud.datastore.Blob;
+import com.google.cloud.datastore.BlobValue;
+import com.google.cloud.datastore.BooleanValue;
+import com.google.cloud.datastore.DoubleValue;
+import com.google.cloud.datastore.EntityValue;
+import com.google.cloud.datastore.FullEntity;
+import com.google.cloud.datastore.IncompleteKey;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.LongValue;
+import com.google.cloud.datastore.NullValue;
+import com.google.cloud.datastore.PathElement;
+import com.google.cloud.datastore.StringValue;
+import com.google.cloud.datastore.TimestampValue;
+import com.google.cloud.datastore.Value;
+import com.google.common.base.Strings;
+import org.apache.hadoop.io.NullWritable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link BatchSink} that writes data to Cloud Datastore.
+ * This {@link DatastoreSink} takes a {@link StructuredRecord} in, converts it to Entity, and writes it to the
+ * Cloud Datastore kind.
+ */
+@Plugin(type = BatchSink.PLUGIN_TYPE)
+@Name(DatastoreSink.PLUGIN_NAME)
+@Description("CDAP Google Cloud Datastore Batch Sink takes the structured record from the input source and writes "
+  + "the data into Google Cloud Datastore.")
+public class DatastoreSink extends BatchSink<StructuredRecord, NullWritable, FullEntity<?>> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DatastoreSink.class);
+
+  public static final String PLUGIN_NAME = "Datastore";
+
+  private final DatastoreSinkConfig config;
+
+  public DatastoreSink(DatastoreSinkConfig config) {
+    this.config = config;
+  }
+
+  @SuppressWarnings("ConstantConditions")
+  @Override
+  public void prepareRun(BatchSinkContext context) {
+    Schema inputSchema = context.getInputSchema();
+    LOG.debug("DatastoreSink `prepareRun` input schema: {}", inputSchema);
+    config.validate(inputSchema);
+
+    context.addOutput(Output.of(config.getReferenceName(), new DatastoreOutputFormatProvider(config)));
+
+    LineageRecorder lineageRecorder = new LineageRecorder(context, config.getReferenceName());
+    lineageRecorder.createExternalDataset(inputSchema);
+    // Record the field level WriteOperation
+    lineageRecorder.recordWrite("Write", "Wrote to Datastore sink",
+                                inputSchema.getFields().stream()
```

Comment:
Auto-generated keys are visible in the cloud console since users see them after records have been inserted.
In our case when we insert records to the Datastore, we don't have auto-generated keys upfront, they are generated during insert internally by Datastore API.

I have tried to add auto-generated key to the schema (updated incoming schema):
```
LineageRecorder lineageRecorder = new LineageRecorder(context, config.getReferenceName());
    Schema lineageSchema = config.shouldUseAutoGeneratedKey()
      ? Schema.recordOf(context.getInputSchema().getRecordName(), Stream.concat(context.getInputSchema().getFields().stream(),
                                                                   Stream.of(Schema.Field.of(config.getKeyAlias(), Schema.of(Schema.Type.LONG))))
      .toArray(Schema.Field[]::new))
      : context.getInputSchema();
    lineageRecorder.createExternalDataset(lineageSchema);
    // Record the field level WriteOperation
    lineageRecorder.recordWrite("Write", "Wrote to Cloud Datastore sink",
                                lineageSchema.getFields().stream()
                                  .map(Schema.Field::getName)
                                  .collect(Collectors.toList()));
```
Field lineage stopped working after such changes. Log shows the following:
```
2019-02-22 18:01:54,820 - DEBUG [WorkflowDriver:c.c.c.d.SmartWorkflow@658] - Inputs of following operations are neither part of the input schema of a stage nor are generated by any previous operations recorded by that stage: <stage:Google Cloud Datastore2, [operation:Write, field:__key__]>.
```
Looks like CDAP lineage does not support schema change in batch write plugin.